driver/random.hpp

```diff
@@ -1,32 +1,93 @@
 #ifndef GUARD_RANDOM_GEN_
 #define GUARD_RANDOM_GEN_
 
+#include <miopen/env.hpp>
+#include <iostream>
 #include <random>
 
-std::minstd_rand& get_minstd_gen()
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_DRIVER_PRNG_SEED)
+namespace prng {
+namespace details {
+using glibc_gen = std::linear_congruential_engine<std::uint32_t, 1103515245, 12345, 2147483648>;
+
+inline std::random_device::result_type get_default_seed()
+{
+    static std::random_device::result_type seed{[] {
+        auto external_seed = miopen::Value(MIOPEN_DEBUG_DRIVER_PRNG_SEED{}, 100500);
+
+        auto seed = external_seed == 0
+                        ? std::random_device{}()
+                        : static_cast<std::random_device::result_type>(external_seed);
+        std::cout << "PRNG seed: " << seed << "\n";
+        return seed;
+    }()};
+    return seed;
+}
+
+inline glibc_gen& get_prng()
+{
+    static thread_local glibc_gen gen{get_default_seed()};
+    return gen;
+}
+} // namespace details
+
+inline void reset_seed(std::random_device::result_type seed = 0)
 {
-    static thread_local std::minstd_rand minstd_gen(std::random_device{}());
-    return minstd_gen;
+    details::get_prng().seed(seed + details::get_default_seed());
 }
 
+// similar to std::generate_canonical, but simpler and faster
 template <typename T>
-inline T FRAND()
+inline T gen_canonical()
 {
-    double d = std::generate_canonical<double, 1>(get_minstd_gen());
-    return static_cast<T>(d);
+    if constexpr(std::is_floating_point_v<T>) // native fp
+    {
+        static constexpr T range =
+            static_cast<T>(1) /
+            static_cast<T>(details::glibc_gen::max() - details::glibc_gen::min() + 1);
+        return range * static_cast<T>(details::get_prng()() - details::glibc_gen::min());
+    }
+    else if constexpr(std::is_integral_v<T>)
+    {
+        auto val = details::get_prng()();
+        return static_cast<T>(((val >> 4) ^ (val >> 16)) & 0x1);
+    }
+    else
+    {
+        return static_cast<T>(gen_canonical<float>());
+    }
 }
 
-inline int GET_RAND()
+template <typename T>
+inline T gen_0_to_B(T B)
 {
-    decltype(auto) minstd_gen = get_minstd_gen();
-    return minstd_gen();
+    if constexpr(std::is_floating_point_v<T>) // native fp
+    {
+        return gen_canonical<T>() * B;
+    }
+    else if constexpr(std::is_integral_v<T>)
+    {
+        // can only generate 27bit range, so it may not be suitable
+        // for huge 64 bit ranges, but we do not expect such ranges
```

Comment:
[R] static_assert here if sizeof(T) > 4?

Reply:
Yes, but not exactly. It actually depends on B. If B < 2^27 everything is fine even for 64 bit type, but I'd prefer to do not perform runtime check for it.

Reply:
[resolved]